crypto/cipher_extra/aead_test.cc

```diff
@@ -1377,3 +1359,182 @@ TEST(AEADTest, TestGCMSIV256Change16Alignment) {
+
+  for (evp_aead_direction_t direction : directions) {
+    bssl::ScopedEVP_AEAD_CTX ctx;
+    bssl::ScopedEVP_AEAD_CTX ctx2;
+
+    ASSERT_TRUE(EVP_AEAD_CTX_init_with_direction(ctx.get(), params.cipher,
+                                                 params.key, params.key_len,
+                                                 params.tag_len, direction));
+    ASSERT_TRUE(EVP_AEAD_CTX_init_with_direction(ctx2.get(), params.cipher,
+                                                 params.key, params.key_len,
+                                                 params.tag_len, direction));
+
+    bssl::ScopedCBB cbb;
+    CBB_init(cbb.get(), 1024);
+    CBS cbs;
+
+    ASSERT_TRUE(EVP_AEAD_CTX_serialize_state(ctx.get(), cbb.get()));
+    CBS_init(&cbs, CBB_data(cbb.get()), CBB_len(cbb.get()));
+    ASSERT_TRUE(EVP_AEAD_CTX_deserialize_state(ctx2.get(), &cbs));
+  }
+}
+
+TEST_P(EvpAeadCtxSerdeTest, FailUnknownSerdeVersion) {
+  // A minimal DER encoding with the serialization version set to 42.
+  // SEQUENCE {
+  //   INTEGER { 42 }
+  // }
+  static const size_t INVALID_VERSION_DER_LEN = 5;
+  static const uint8_t INVALID_VERSION_DER[INVALID_VERSION_DER_LEN] = {
+      0x30, 0x03, 0x02, 0x01, 0x2a};
+
+  const ParamType &params = GetParam();
+
+  const evp_aead_direction_t directions[] = {evp_aead_open, evp_aead_seal};
+
+  for (const evp_aead_direction_t direction : directions) {
+    bssl::ScopedEVP_AEAD_CTX ctx;
+
+    ASSERT_TRUE(EVP_AEAD_CTX_init_with_direction(ctx.get(), params.cipher,
+                                                 params.key, params.key_len,
+                                                 params.tag_len, direction));
+
+    CBS cbs;
+    CBS_init(&cbs, INVALID_VERSION_DER, INVALID_VERSION_DER_LEN);
+    ASSERT_FALSE(EVP_AEAD_CTX_deserialize_state(ctx.get(), &cbs));
+  }
+}
+
+TEST_P(EvpAeadCtxSerdeTest, FailUnknownCipherId) {
+  // A minimal DER encoding with the serialization version set to 1,
+  // and the cipher identifer set to 65535.
+  // SEQUENCE {
+  //   INTEGER { 1 }
+  //   INTEGER { 65535 }
+  // }
+  static const size_t INVALID_CIPHER_ID_DER_LEN = 10;
+  static const uint8_t INVALID_CIPHER_ID_DER[INVALID_CIPHER_ID_DER_LEN] = {
+      0x30, 0x08, 0x02, 0x01, 0x01, 0x02, 0x03, 0x00, 0xff, 0xff};
+
+  const ParamType &params = GetParam();
+
+  const evp_aead_direction_t directions[] = {evp_aead_open, evp_aead_seal};
+
+  for (const evp_aead_direction_t direction : directions) {
+    bssl::ScopedEVP_AEAD_CTX ctx;
+
+    ASSERT_TRUE(EVP_AEAD_CTX_init_with_direction(ctx.get(), params.cipher,
+                                                 params.key, params.key_len,
+                                                 params.tag_len, direction));
+
+    CBS cbs;
+    CBS_init(&cbs, INVALID_CIPHER_ID_DER, INVALID_CIPHER_ID_DER_LEN);
+    ASSERT_FALSE(EVP_AEAD_CTX_deserialize_state(ctx.get(), &cbs));
+  }
+}
+
+TEST(EvpAeadCtxSerdeTest, ID) {
+  bool identifiers[AEAD_MAX_ID + 1] = {false};
```

Comment:
What does this identifier do/track? I don't quite understand the logic of when this flips to true/false.

Reply:
What this is doing is creating an array of boolean values equal to the number of AEAD cipher identifiers (determined by the AEAD_MAX_ID which is the highest value currently), and then add one so that we can use `identifiers[0]` to be the unknown identifier which should never be used by an AEAD cipher. We then loop over all the test cases and record `identifiers[EVP_AEAD_CTX_get_aead_id()] = true` If an identifier is already been seen then this test case will fail as that is unexpected. At the end we should expect that we tested all the AEAD cipher identifiers (`identifiers[1] ... identifiers[AEAD_MAX_ID]` must all be `true`), and that `identifiers[0] == false`.

This ensures that:
1. All identifiers have a uniquely assigned identifier number
2. No AEAD algorithm returns the default value `0` aka the `AEAD_UNKNOWN_ID` value.
3. As new AEAD identifiers are added (and thus AEAD_MAX_ID is updated) that they are added to the test matrix here and are not missed in testing.